src/Tasks/MapSourceRoots.cs

```diff
@@ -0,0 +1,117 @@
+    {
+        /// <summary>
+        /// SourceRoot items with the following optional well-known metadata:
+        /// <list type="bullet">
+        ///   <term>SourceControl</term><description>Indicates name of the source control system the source root is tracked by (e.g. Git, TFVC, etc.), if any.</description>
+        ///   <term>NestedRoot</term><description>If a value is specified the source root is nested (e.g. git submodule). The value is a path to this root relative to the containing root.</description>
+        ///   <term>ContainingRoot</term><description>Identifies another source root item that this source root is nested under.</description>
+        /// </list>
+        /// </summary>
+        [Required]
+        public ITaskItem[] SourceRoots { get; set; }
+
+        [Output]
+        public ITaskItem[] MappedSourceRoots { get; private set; }
+
+        private static class Names
+        {
+            public const string SourceRoot = nameof(SourceRoot);
+
+            // Names of well-known SourceRoot metadata items:
+            public const string SourceControl = nameof(SourceControl);
+            public const string NestedRoot = nameof(NestedRoot);
+            public const string ContainingRoot = nameof(ContainingRoot);
+            public const string MappedPath = nameof(MappedPath);
+        }
+
+        public override bool Execute()
+        {
+            var topLevelMappedPaths = new Dictionary<string, string>();
+            bool success = true;
+            int i = 0;
+
+            void SetTopLevelMappedPaths(bool sourceControl)
+            {
+                foreach (var root in SourceRoots)
+                {
+                    if (!string.IsNullOrEmpty(root.GetMetadata(Names.SourceControl)) == sourceControl)
+                    {
+                        string nestedRoot = root.GetMetadata(Names.NestedRoot);
+                        if (string.IsNullOrEmpty(nestedRoot))
+                        {
+                            if (topLevelMappedPaths.ContainsKey(root.ItemSpec))
+                            {
+                                Log.LogErrorFromResources("MapSourceRoots.ContainsDuplicate", Names.SourceRoot, root.ItemSpec);
+                                success = false;
+                            }
+                            else
+                            {
+                                var mappedPath = "/_" + (i == 0 ? "" : i.ToString()) + "/";
+                                topLevelMappedPaths.Add(root.ItemSpec, mappedPath);
+                                root.SetMetadata(Names.MappedPath, mappedPath);
+                                i++;
+                            }
+                        }
+                    }
+                }
+            }
+
+            string EndWithSlash(string path)
+                => (path[path.Length - 1] == '/') ? path : path + '/';
+
+            // assign mapped paths to process source control roots first:
```

Comment:
What's the intention behind this ordering?

Reply:
In most cases there will be a single SourceRoot that corresponds to the repository root, plus some additional SourceRoots for source packages and other external source roots.

The first SourceRoot in the item group is mapped to `/_/` path prefix. Since the mapped paths are visible to the user it's better if `/_/` prefix is commonly used for all the files in the repo (as opposed to `/_N/` prefix based on the order the targets added their SourceRoots).

I'll add a comment.